File: Yoonseo Jeong/2023-05-06-valid-parentheses.js
var isValid = function(s) {
    if(!s || s.length < 1) return false
    
    // check if the string length is even number
    // if not, it means one of them doesn't have pair
    if(pair.length % 2 === 1) return false

    // Stack - to push and pop in the correnct order to match items
    const pair = []
    
    // loop through the string char
    for (let i = 0; i < s.length; i++) {
        // if char is opening braket, push pair closing braket to array(stack)
        if (s[i] === "(") {
        	pair.push(")")
        }
        else if (s[i] === "{") {
        	pair.push("}")
        }
        else if (s[i] === "[") {
        	pair.push("]")
        }
        // if char is closing braket, compare with last item of pair array
        // if not same, then return false
        else if(s[i] !== pair.pop()) {
        	return false
        }
    }
    return true

    // complexity
    // time: O(n)
    // space: O(n)
};

File: Mave_Hur/2023-05-06-Valid-Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        paren_dict = {")":"(","}":"{","]":"["}
        stack = []

        for char in s:
            # char is closing parentheses
            if char in paren_dict:
                if not stack or stack[-1] != paren_dict[char]:
                    return False
                else:
                    stack.pop()

            # char is opening parentheses
            else:
                stack.append(char)

        return not stack

    # Time complexity: o(n)
    # Space complexity: o(n)
    